test/spinxsk/spinxsk.c

```diff
@@ -1992,6 +2022,182 @@ ProcessPkts(
+        } else if (frameType == 1) {
+            //
+            // Build a TCP frame.
+            // Up to 40 bytes of TCP options and random flags.
+            //
+            UINT8 tcpOptions[TCP_MAX_OPTION_LEN];
+            FillRandomBuffer(sizeof(tcpOptions), tcpOptions);
+            UINT16 tcpOptionsLength = 4 * (RandUlong() % (10 + 1));
+            const UINT32 thSeq = RandUlong();
+            const UINT32 thAck = RandUlong();
+            const UINT8 thFlags = (UINT8)RandUlong();
+            const UINT16 thWin = (UINT16)RandUlong();
+
+            ASSERT_FRE(PktBuildTcpFrame(
+                &frame[backfill], &frameLength, payload, (UINT16)payloadLength,
+                tcpOptions, tcpOptionsLength, thSeq, thAck, thFlags, thWin,
+                &localMac, &remoteMac, af, &localIp, &remoteIp, localPort, remotePort));
+        } else {
+            //
+            // Send pure chaos.
+            //
+            frameLength = RandUlong() % (frameLength + 1);
+            for (UINT32 j = 0; j + sizeof(ULONG) < frameLength; j += sizeof(ULONG)) {
+                *(ULONG*)(frame + backfill + j) = RandUlong();
+            }
+        }
+
+        //
+        // Setup the fnmp data frame.
+        //
+        DATA_FRAME fnmpFrame = {0};
+        fnmpFrame.Input.RssHashQueueId = RandUlong() % queueCount;
+        fnmpFrame.Input.Checksum.Value = (PVOID)(ULONG_PTR)RandUlong();
+        fnmpFrame.Input.Rsc.Value = (PVOID)(ULONG_PTR)RandUlong();
+
+        //
+        // Split the frame into one or two MDLs.
+        //
+        DATA_BUFFER buffers[2] = {0};
+        if (RandUlong() % 2) {
+            buffers[0].DataOffset = backfill;
+            buffers[0].DataLength = frameLength;
+            buffers[0].BufferLength = sizeof(frame);
+            buffers[0].VirtualAddress = frame;
+            fnmpFrame.Buffers = buffers;
+            fnmpFrame.BufferCount = 1;
+        } else {
+            const UINT32 bufferSplitOffset = RandUlong() % (frameLength + 1);
+
+            buffers[0].DataOffset = backfill;
+            buffers[0].DataLength = bufferSplitOffset;
+            buffers[0].BufferLength = backfill + buffers[0].DataLength;
+            buffers[0].VirtualAddress = frame;
+            buffers[1].DataLength = frameLength - bufferSplitOffset;
+            buffers[1].BufferLength = sizeof(frame) - buffers[0].BufferLength;
+            buffers[1].VirtualAddress = frame + buffers[0].BufferLength;
+            fnmpFrame.Buffers = buffers;
+            fnmpFrame.BufferCount = 2;
+        }
+
+        FnMpRxEnqueue(Fnmp, &fnmpFrame);
+    }
+
+    DATA_FLUSH_OPTIONS flushOptions = {0};
+    flushOptions.Flags.DpcLevel = RandUlong() & 1;
+    flushOptions.Flags.LowResources = RandUlong() & 1;
+    flushOptions.Flags.RssCpu = RandUlong() & 1;
+    flushOptions.RssCpuQueueId = RandUlong() % queueCount;
+
+    FnMpRxFlush(Fnmp, &flushOptions);
+
+    //
+    // Cleanup
+    //
+    free(payload);
```

Comment:
nit: generally we try to clean up in reverse order things are declared/allocated. obviously irrelevant in practice here, but it's a great pattern to follow in all cases, because then a future PR need not reorder these if an intermediate cleanup dependency is introduced.